src/main/ptapi/io/piuio/lxio/lxio.c

```diff
@@ -0,0 +1,296 @@
+      ptapi_piuio_pad_in[i][j].rd = lxio_btn_state_buffer[i * 4 + j] & (1 << 4);
+    }
+  }
+
+  // cabinet buttons
+  ptapi_piuio_sys.test = lxio_btn_state_buffer[8] & (1 << 1);
+  ptapi_piuio_sys.service = lxio_btn_state_buffer[8] & (1 << 6);
+  ptapi_piuio_sys.clear = lxio_btn_state_buffer[8] & (1 << 7);
+  ptapi_piuio_sys.coin = lxio_btn_state_buffer[8] & (1 << 2);
+  ptapi_piuio_sys.coin2 = lxio_btn_state_buffer[9] & (1 << 2);
+
+}
+
+bool ptapi_io_piuio_recv(void)
+{
+  // NOTE: The lxio *MUST* be written to in order to be read.
+  // It cannot be read individually without a written payload. It will soft
+  // lock. In order to ensure the full cycle, everything is taken care of here.
+
+  // take the ptapi info and convert it to an lxio message.
+  convert_output_to_lxio();
+
+  if (!lxio_drv_device_write(lxio_light_state_buffer, LXIO_MSG_SIZE)) {
+    return false;
+  }
+
+  if (!lxio_drv_device_read(lxio_btn_state_buffer, LXIO_MSG_SIZE)) {
+    return false;
+  }
+
+  // TODO: lxio contains an unsigned 16bit packet counter in the last two LSB
+  // bytes, we can check that for errors....or not. libusb will tell us.
+
+  // input from lxio is active low, so invert the recv data for active high
+  // logic in ptapi.
+  for (uint8_t j = 0; j < LXIO_MSG_SIZE; j++) {
+    lxio_btn_state_buffer[j] ^= 0xFF;
+  }
+
+  // take the lxio message and convert it to ptapi.
+  convert_input_to_piuio();
+
+  return true;
+}
+
+bool ptapi_io_piuio_send(void)
+{
+  // taken care of in recv
+  return true;
+}
+
+void ptapi_io_piuio_get_input_pad(
+    uint8_t player,
+    enum ptapi_io_piuio_sensor_group sensor_group,
+    struct ptapi_io_piuio_pad_inputs *inputs)
+{
+  memcpy(
+      inputs,
+      &ptapi_piuio_pad_in[player][sensor_group],
+      sizeof(struct ptapi_io_piuio_pad_inputs));
+}
+
+void ptapi_io_piuio_get_input_sys(struct ptapi_io_piuio_sys_inputs *inputs)
+{
+  memcpy(inputs, &ptapi_piuio_sys, sizeof(struct ptapi_io_piuio_sys_inputs));
+}
+
+void ptapi_io_piuio_set_output_pad(
+    uint8_t player, const struct ptapi_io_piuio_pad_outputs *outputs)
+{
+  memcpy(
+      &ptapi_piuio_pad_out[player],
+      outputs,
+      sizeof(struct ptapi_io_piuio_pad_outputs));
+}
+
+void ptapi_io_piuio_set_output_cab(
+    const struct ptapi_io_piuio_cab_outputs *outputs)
+{
+  memcpy(
+      &ptapi_piuio_cab_out, outputs, sizeof(struct ptapi_io_piuio_cab_outputs));
+}
+
+
+const char* ptapi_io_piubtn_ident(void)
+{
+    return "lxio";
+}
+
+bool ptapi_io_piubtn_open(void)
+{
+  memset(lxio_btn_state_buffer, 0, sizeof(lxio_btn_state_buffer));
+  return lxio_drv_device_open();
+}
+
+void ptapi_io_piubtn_close(void)
+{
+  lxio_drv_device_close();
+}
+
+bool ptapi_io_piubtn_recv(void)
+{
+    ptapi_io_piuio_recv();
+    return true;
+}
+
+bool ptapi_io_piubtn_send(void)
+{
+    return true;
+}
+
+void ptapi_io_piubtn_get_input(uint8_t player, struct ptapi_io_piubtn_inputs* inputs)
+{
+    // Set all inputs to off
+    if (player == 0) {
```

Comment:
Hm, considering the threading situation mentioned, this part could run into consistency issues if the two IOs are threaded by separate threads. Locking can be quite expensive and would also need to be done in the `ptapi_io_piuio_recv` which is pumping the data. 

We can call it fine for now, but I would definitely leave a comment about the decision to "take the risk" now that there might be lost updates with the buttons. the pad inputs should not experience that if the IO is driven from there (which is the case right now).